source/common/upstream/outlier_detection_impl.cc

```diff
@@ -92,6 +93,31 @@ void DetectorHostMonitorImpl::putHttpResponseCode(uint64_t response_code) {
     consecutive_5xx_ = 0;
     consecutive_gateway_failure_ = 0;
   }
+
+  if (monitors_set_.empty()) {
+    return;
+  }
+  // Wrap reported HTTP code into outlier detection extension's wrapper and forward
+  // it to configured extensions.
+  monitors_set_.forEach([response_code](const ExtMonitorPtr& monitor) {
+    monitor->putResult(HttpCode(response_code));
+  });
+}
+ExtMonitor::ExtMonitorCallback DetectorHostMonitorImpl::getOnFailedExtensioMonitorCallback() {
+  return [this](uint32_t enforce, std::string failed_monitor_name,
+                absl::optional<std::string> extra_info) {
+    std::shared_ptr<DetectorImpl> detector = detector_.lock();
+    if (!detector) {
+      // It's possible for the cluster/detector to go away while we still have a host in use.
+      return;
+    }
+    failed_monitor_name_ = failed_monitor_name;
+    if (extra_info.has_value()) {
+      failed_extra_info_ = extra_info.value();
+    }
+    failed_monitor_enforce_ = enforce;
+    detector->notifyMainThreadConsecutiveError(host_.lock(), envoy::data::cluster::v3::EXTENSION);
+  };
```

Comment:
Yeah. That is a good point. Let me think about changing API to get that info.

Reply:
Refactored `notifyMainThreadConsecutiveError` method to take the name of `failed_monitor` and `enforce` value and pass it to the main thread, instead of main thread querying detector for those values (failed_monitor_name and enforce). Calling `notifyMainThreadConsecutiveError` is thread safe and if more than one monitor reports failure, several events will be posted to the main thread and will be processed sequentially.